### PR TITLE
refactor: add usecases for createvoteservice

### LINKED ISSUE #1344: [REFACTOR]: add useCases for createVoteService
## Description

Extract useCases from the create vote service and then add tests for the useCases.

### PR DESCRIPTION
**Pull request by CatiaAntunes96**
Relates to:
- #1344 

## Proposed Changes

  - Add useCases for creating a vote on a cardItem and on a cardGroup
  - Add tests for both useCases
  - Refactor tests on create vote service

This pull request closes #1344

### CODE REVIEW

File: backend/src/modules/votes/interfaces/services/create.vote.service.interface.ts [hunk 1]
```diff
@@ -1,18 +1,17 @@
...
-		count: number,
-		retryCount?: number
-	): Promise<void>;
+	// addVoteToCardGroup(
+	// 	boardId: string,
+	// 	cardId: string,
+	// 	userId: string,
+	// 	count: number,
+	// 	retryCount?: number
+	// ): Promise<void>;
```
Comment (CatiaBarroco-xgeeks) [inline backend/src/modules/votes/interfaces/services/create.vote.service.interface.ts:16]: this is necessary?

File: backend/src/modules/votes/services/create.vote.service.spec.ts [hunk 2]
```diff
@@ -79,116 +74,52 @@ describe('CreateVoteService', () => {
...
-			//if updateBoardUserServiceMock.updateVoteUser fails
-			try {
-				updateBoardUserServiceMock.updateVoteUser.mockResolvedValueOnce(null);
-				await voteService.addVoteToCard(board._id, card._id, userId, cardItem._id, 1);
-			} catch (ex) {
-				expect(ex).toBeInstanceOf(InsertFailedException);
-			}
+		it('should return if user can vote', async () => {
+			const board_1 = { ...board };
+			board_1.maxVotes = null;
```
Comment (CatiaBarroco-xgeeks) [inline backend/src/modules/votes/services/create.vote.service.spec.ts:106]: This name is not a good practice, when we want something specific like "1" the best thing is to put "one",
but what is the purpose of this constant?

File: backend/src/modules/votes/dto/useCase/create-card-group-vote.use-case.dto.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	@IsString()
+	userId: string;
+
+	@IsNotEmpty()
+	@IsNumber()
+	count: number;
+
+	@IsOptional()
+	@IsNumber()
+	retryCount?: number;
```
Comment (CatiaBarroco-xgeeks) [inline backend/src/modules/votes/dto/useCase/create-card-group-vote.use-case.dto.ts:22]: I'm not sure, can you confirm?
if you have the `@Optional` it is necessary the `"?"`

File: backend/src/modules/votes/interfaces/services/create.vote.service.interface.ts (hunk 1 above)
Comment (CatiaBarroco-xgeeks) [reply backend/src/modules/votes/interfaces/services/create.vote.service.interface.ts]: Done https://github.com/xgeekshq/split/pull/1354/commits/db0ab91a74cd3dac4294522d8f7116c295cd65dd

File: backend/src/modules/votes/services/create.vote.service.spec.ts (hunk 2 above)
Comment (CatiaAntunes96) [reply backend/src/modules/votes/services/create.vote.service.spec.ts:106]: In order to not change the original structure of the variable board, which is created before starting the tests, when I need to change that object, I create a copy of the variable. Thus, it does not affect other tests or structures that use the board constant.

File: backend/src/modules/votes/dto/useCase/create-card-group-vote.use-case.dto.ts (hunk 3 above)
Comment (CatiaAntunes96) [reply backend/src/modules/votes/dto/useCase/create-card-group-vote.use-case.dto.ts:22]: Yes, the decorator evaluates the validation of the property of the class, but when you try to access an Object of that type, the typescript needs to know that that property can be undefined.

File: backend/src/modules/votes/services/create.vote.service.spec.ts (hunk 2 above)
Comment (CatiaBarroco-xgeeks) [reply backend/src/modules/votes/services/create.vote.service.spec.ts]: try something like"boardOne" without the "_"
but I still think it is a strange name. can't we come up with a better name for what that constant is doing?

File: backend/src/modules/votes/applications/create/create-card-group-vote.use-case.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+@Injectable()
+export class CreateCardGroupVoteUseCase implements UseCase<CreateCardGroupVoteUseCaseDto, void> {
+	private logger: Logger = new Logger('CreateVoteService');
+	constructor(
+		@Inject(TYPES.services.CreateVoteService)
+		private readonly createVoteService: CreateVoteServiceInterface,
+		@Inject(TYPES.repositories.VoteRepository)
+		private readonly voteRepository: VoteRepositoryInterface,
+		@Inject(BoardUsers.TYPES.services.UpdateBoardUserService)
+		private updateBoardUserService: UpdateBoardUserServiceInterface
```
Comment (nunocaseiro) [inline backend/src/modules/votes/applications/create/create-card-group-vote.use-case.ts:22]:
```suggestion
		private readonly updateBoardUserService: UpdateBoardUserServiceInterface
```

File: backend/src/modules/votes/applications/create/create-card-item-vote.use-case.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+@Injectable()
+export class CreateCardItemVoteUseCase implements UseCase<CreateCardItemVoteUseCaseDto, void> {
+	private logger: Logger = new Logger('CreateVoteService');
+	constructor(
+		@Inject(TYPES.services.CreateVoteService)
+		private readonly createVoteService: CreateVoteServiceInterface,
+		@Inject(TYPES.repositories.VoteRepository)
+		private readonly voteRepository: VoteRepositoryInterface,
+		@Inject(BoardUsers.TYPES.services.UpdateBoardUserService)
+		private updateBoardUserService: UpdateBoardUserServiceInterface
```
Comment (nunocaseiro) [inline backend/src/modules/votes/applications/create/create-card-item-vote.use-case.ts:22]:
```suggestion
		private readonly updateBoardUserService: UpdateBoardUserServiceInterface
```

File: backend/src/modules/votes/controller/votes.controller.ts [hunk 6]
```diff
@@ -21,20 +21,24 @@ import { UnauthorizedResponse } from 'src/libs/swagger/errors/unauthorized.swagg
...
 	constructor(
-		@Inject(TYPES.applications.CreateVoteApplication)
-		private createVoteApp: CreateVoteApplicationInterface,
 		@Inject(TYPES.applications.DeleteVoteApplication)
 		private deleteVoteApp: DeleteVoteApplicationInterface,
+		@Inject(TYPES.applications.CreateCardItemVoteUseCase)
+		private createCardItemVoteUseCase: UseCase<CreateCardItemVoteUseCaseDto, void>,
+		@Inject(TYPES.applications.CreateCardGroupVoteUseCase)
+		private createCardGroupVoteUseCase: UseCase<CreateCardGroupVoteUseCaseDto, void>,
 		private socketService: SocketGateway
```
Comment (nunocaseiro) [inline backend/src/modules/votes/controller/votes.controller.ts:42]: Add `readonly`

File: backend/src/modules/votes/controller/votes.controller.ts [hunk 7]
```diff
@@ -75,7 +79,13 @@ export default class VotesController {
...
+			await this.createCardItemVoteUseCase.execute({
+				boardId,
+				cardId,
+				userId: request.user._id,
+				cardItemId: itemId,
+				count
+			});
 		}
 
 		this.socketService.sendUpdateVotes(socketId, data);
```
Comment (nunocaseiro) [inline backend/src/modules/votes/controller/votes.controller.ts:91]: This should be moved to the use-case. There it should decide which operation should be made

File: backend/src/modules/votes/controller/votes.controller.ts [hunk 8]
```diff
@@ -119,7 +129,12 @@ export default class VotesController {
...
-			await this.createVoteApp.addVoteToCardGroup(boardId, cardId, request.user._id, count);
+			await this.createCardGroupVoteUseCase.execute({
+				boardId,
+				cardId,
+				userId: request.user._id,
+				count
+			});
 		}
 
 		this.socketService.sendUpdateVotes(socketId, data);
```
Comment (nunocaseiro) [inline backend/src/modules/votes/controller/votes.controller.ts:140]: same as above. You can create a `card-item-vote.use-case.ts` and a `card-group-vote.use-case.ts`

File: backend/src/modules/votes/services/create.vote.service.ts [hunk 9]
```diff
@@ -15,83 +12,35 @@ import { GetBoardServiceInterface } from 'src/modules/boards/interfaces/services
...
 export default class CreateVoteService implements CreateVoteServiceInterface {
 	constructor(
-		@Inject(TYPES.repositories.VoteRepository)
-		private readonly voteRepository: VoteRepositoryInterface,
 		@Inject(BoardUsers.TYPES.services.GetBoardUserService)
 		private getBoardUserService: GetBoardUserServiceInterface,
 		@Inject(BoardUsers.TYPES.services.UpdateBoardUserService)
 		private updateBoardUserService: UpdateBoardUserServiceInterface,
 		@Inject(Boards.TYPES.services.GetBoardService)
 		private getBoardService: GetBoardServiceInterface
```
Comment (nunocaseiro) [inline backend/src/modules/votes/services/create.vote.service.ts:20]: Add readonly

File: backend/src/modules/votes/services/create.vote.service.ts [hunk 10]
```diff
@@ -15,83 +12,35 @@ import { GetBoardServiceInterface } from 'src/modules/boards/interfaces/services
...
-
-	/* #################### HELPERS #################### */
-
-	private async canUserVote(boardId: string, userId: string, count: number) {
+	async canUserVote(boardId: string, userId: string, count: number) {
 		const canUserVoteResult = await this.verifyIfUserCanVote(boardId, userId, count);
 
 		if (!canUserVoteResult) throw new InsertFailedException(INSERT_VOTE_FAILED);
 
 		return;
```
Comment (nunocaseiro) [inline backend/src/modules/votes/services/create.vote.service.ts:28]: I think this line isn't need

**Review by nunocaseiro — CHANGES_REQUESTED** [review]

File: backend/src/modules/votes/controller/votes.controller.ts (hunk 7 above)
Comment (CatiaAntunes96) [reply backend/src/modules/votes/controller/votes.controller.ts:91]: I agree, but I was thinking of doing that when the changes that are in another PR (#1351) are approved. But I will change the names to aggregate the logic of the card item and card groups

File: backend/src/modules/votes/services/create.vote.service.spec.ts (hunk 2 above)
Comment (CatiaBarroco-xgeeks) [reply backend/src/modules/votes/services/create.vote.service.spec.ts]: Done fb79a164f71c20ef1b74ee978bad19a4107bd539

**Review by nunocaseiro — APPROVED** [review]
LGTM 🚀

**Pull request merged**